src/metabase/lib/column_group.cljc

```diff
@@ -39,59 +50,96 @@
-        ;; stage display name.
-        {:display-name (lib.metadata.calculation/display-name query stage-number stage)}))
-     {:is-from-join           false
-      :is-implicitly-joinable false})
-
-    :group-type/join.explicit
-    (merge
-     (let [join-alias (:join-alias column-group)]
-       (when-let [join (lib.join/resolve-join query stage-number join-alias)]
-         (lib.metadata.calculation/display-info query stage-number join)))
-     {:is-from-join           true
-      :is-implicitly-joinable false})
-
-    :group-type/join.implicit
-    (merge
-     (let [fk-field-id (:fk-field-id column-group)]
-       (when-let [field (lib.metadata/field query fk-field-id)]
-         (let [field-info (lib.metadata.calculation/display-info query stage-number field)]
-           ;; Implicitly joined column pickers don't use the target table's name, they use the FK field's name with
-           ;; "ID" dropped instead.
-           ;; This is very intentional: one table might have several FKs to one foreign table, each with different
-           ;; meaning (eg. ORDERS.customer_id vs. ORDERS.supplier_id both linking to a PEOPLE table).
-           ;; See #30109 for more details.
-           (assoc field-info :fk-reference-name (lib.util/strip-id (:display-name field-info))))))
-     {:is-from-join           false
-      :is-implicitly-joinable true})))
+  (display-info-for-group-method query stage-number column-group))
 
-(mu/defn ^:private column-group-info :- [:map [::group-type GroupType]]
-  "The value we should use to `group-by` inside [[group-columns]]."
-  [{source :lib/source, :as column-metadata} :- lib.metadata/ColumnMetadata]
-  (case source
-    :source/implicitly-joinable
-    {::group-type :group-type/join.implicit, :fk-field-id (:fk-field-id column-metadata)}
+(defmulti ^:private column-group-info-method
+  {:arglists '([column-metadata])}
+  :lib/source)
+
+(defmethod column-group-info-method :source/implicitly-joinable
+  [column-metadata]
+  {::group-type :group-type/join.implicit, :fk-field-id (:fk-field-id column-metadata)})
+
+(defmethod column-group-info-method :source/joins
+  [{:keys [table-id], :lib/keys [card-id], :as column-metadata}]
+  (merge
+   {::group-type :group-type/join.explicit}
+   ;; if we're in the process of BUILDING a join and using this in combination
+   ;; with [[metabase.lib.join/join-condition-rhs-columns]], the alias won't be present yet, so group things by the
+   ;; joinable -- either the Card we're joining, or the Table we're joining. Prefer `:lib/card-id` because when we
+   ;; join a Card the Fields might have `:table-id` but we want the entire Card to appear as one group. See #32493
+   (or
+    (when-let [join-alias (lib.join/current-join-alias column-metadata)]
+      {:join-alias join-alias})
+    (when card-id
+      {:card-id card-id})
+    (when table-id
+      {:table-id table-id}))))
```

Comment:
The actual fix -- if we're grouping columns with `:source/joins` and it doesn't have a join alias (which is the case when using `join-condition-rhs-columns` when building a new join), we need to group by the Card ID or Table ID (the ID whatever it is we're joining against) instead. Then we can use this ID to add appropriate display-info for the group.